module/epochs/base_contract_client.go

```diff
@@ -0,0 +1,103 @@
+package epochs
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/rs/zerolog"
+
+	sdk "github.com/onflow/flow-go-sdk"
+	sdkcrypto "github.com/onflow/flow-go-sdk/crypto"
+
+	"github.com/onflow/flow-go/module"
+)
+
+// BaseContractClient represents the core fields and methods needed to create
+// a client to a contract on the Flow Network.
+type BaseContractClient struct {
+	log zerolog.Logger // default logger
+
+	ContractAddress string           // contract address
+	AccountKeyIndex uint             // account key index
+	Signer          sdkcrypto.Signer // signer used to sign transactions
+	Account         *sdk.Account     // account belonging to node interacting with the contract
+
+	Client module.SDKClientWrapper // flow access node client
+}
+
+// NewBaseContractClient ...
+func NewBaseContractClient(log zerolog.Logger,
+	flowClient module.SDKClientWrapper,
+	accountAddress string,
+	accountKeyIndex uint,
+	signer sdkcrypto.Signer,
+	contractAddress string) (*BaseContractClient, error) {
+
+	// get account from access node for given address
+	account, err := flowClient.GetAccount(context.Background(), sdk.HexToAddress(accountAddress))
+	if err != nil {
+		return nil, fmt.Errorf("could not get account: %w", err)
+	}
+
+	// check if account key index within range of keys
+	if len(account.Keys) <= int(accountKeyIndex) {
+		return nil, fmt.Errorf("given account key index is bigger than the number of keys for this account")
+	}
+
+	return &BaseContractClient{
+		ContractAddress: contractAddress,
```

Comment:
`log` field is not set 🤔 . Does a `nil` `zerolog.Logger` not crash, or are we somehow not going down a codepath using it on any tests?

```suggestion
	return &BaseContractClient{
		log: log,
		ContractAddress: contractAddress,
```